bst.py

```diff
@@ -50,14 +51,111 @@ def insert(self, value):
                 child = 'left'
             if traverse is None:
                 #actual insert
-                self.tree[value] = {'depth': depth}
+                self.tree[value] = {}
+
                 self.tree[current][child] = value
+                self.tree[value]['parent'] = current
                 self._size += 1
                 if depth > self._depth:
                     self._depth = depth
                 return
             current = traverse
 
+    def delete(self, val):
+        if self.contains(val):
+            left_child = self.left(val)
+            right_child = self.right(val)
+            parent = self.parent(val)
+            if left_child is None and right_child is None:
+            # Delete a node with not children
+                if parent is not None:
+                # For a node with a parent...
+                    if self.left(parent) == val:
+                        del self.tree[parent]['left']
+                    else:
+                        del self.tree[parent]['right']
+                else:
+                    # For deleting a node that is the root with no children
+                    self.top = None
+                del self.tree[val]
+            elif left_child is not None:
+                biggest_child = self._rightmost(left_child)
+                self._swap_nodes(val, biggest_child)
+                self.delete(val)
+            else:
+                if parent is not None:
+                    if self.left(parent) == val:
+                        self.tree[parent]['left'] = right_child
+                    else:
+                        self.tree[parent]['right'] = right_child
+                    self.tree[right_child]['parent'] = parent
+                else:
+                    self.top = right_child
+                del self.tree[val]
```

Comment:
I see three conditions here:

1.  the node has no children
2.  the node has at least a left child
3.  any other situation

I don't think this is an exhaustive set of the possibilities for deletion.  Make sure you have tests that cover having a child to the left, a child to the right and children on both sides.  Make sure you test having extra nodes below the children of the node to be deleted.  With exhaustive tests you can be sure everything is working, but your tests do not appear exhaustive to me.